x/ccv/provider/types/msg.go

```diff
@@ -185,3 +188,47 @@ func (msg MsgSubmitConsumerMisbehaviour) GetSigners() []sdk.AccAddress {
 	}
 	return []sdk.AccAddress{addr}
 }
+
+func NewMsgSubmitConsumerDoubleVoting(submitter sdk.AccAddress, ev *tmtypes.DuplicateVoteEvidence, header *ibctmtypes.Header) (*MsgSubmitConsumerDoubleVoting, error) {
+	return &MsgSubmitConsumerDoubleVoting{Submitter: submitter.String(), DuplicateVoteEvidence: ev, InfractionBlockHeader: header}, nil
+}
+
+// Route implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerDoubleVoting) Route() string { return RouterKey }
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerDoubleVoting) Type() string {
+	return TypeMsgSubmitConsumerDoubleVoting
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerDoubleVoting) ValidateBasic() error {
+	if msg.Submitter == "" {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, msg.Submitter)
+	}
+	if msg.DuplicateVoteEvidence == nil {
+		return fmt.Errorf("double voting evidence cannot be nil")
+	}
+
+	if msg.InfractionBlockHeader.Header == nil {
+		return fmt.Errorf("infraction header cannot be nil")
+	}
+
+	return nil
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerDoubleVoting) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(&msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgSubmitConsumerDoubleVoting) GetSigners() []sdk.AccAddress {
+	addr, err := sdk.AccAddressFromBech32(msg.Submitter)
+	if err != nil {
+		// same behavior as in cosmos-sdk
+		panic(err)
+	}
+	return []sdk.AccAddress{addr}
```

Comment:
Is my understanding correct that `GetSigners` builds the addresses that need to have signed the `MsgSubmitConsumerDoubleVoting` in order for the message to be valid?
For example, in the [bank module](https://github.com/cosmos/cosmos-sdk/blob/147758fd179f334546d757794509cc2d1ad85b99/x/bank/types/msgs.go#L18), you need to verify that someone sending money from an account `from` has indeed signed the message, otherwise you could potentially steal tokens.

If the above is true, then in the case of `MsgSubmitConsumerDoubleVoting` why do we need the message to be signed by `addr`?

Reply:
Good question. AFAIU all messages must be signed by the submitter and should satisfy the [msg interface ](https://github.com/cosmos/cosmos-sdk/blob/v0.45.16/types/tx_msg.go#L9).